File: taurus-protect-sdk-typescript/src/mappers/reservation.ts
/**
 * Reservation mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { Reservation, ReservationUtxo } from '../models/reservation';
import { safeDate, safeInt, safeMap, safeString } from './base';
import { currencyFromDto } from './currency';

/**
 * Maps a reservation DTO to a Reservation domain model.
 *
 * @param dto - The DTO from the OpenAPI client
 * @returns The Reservation domain model, or undefined if dto is null/undefined
 */
export function reservationFromDto(dto: unknown): Reservation | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    amount: safeString(d.amount),
    creationDate: safeDate(d.creationDate ?? d.creation_date),
    kind: safeString(d.kind),
    comment: safeString(d.comment),
    // Note: OpenAPI DTO uses 'addressid' (lowercase 'i'), map to 'addressId'
    addressId: safeString(d.addressid ?? d.addressId ?? d.address_id),
    address: safeString(d.address),
    currencyInfo: currencyFromDto(d.currencyInfo ?? d.currency_info),
    resourceId: safeString(d.resourceId ?? d.resource_id),
    resourceType: safeString(d.resourceType ?? d.resource_type),
  };
}

/**
 * Maps an array of reservation DTOs to Reservation domain models.
 *
 * @param dtos - Array of DTOs from the OpenAPI client
 * @returns Array of Reservation domain models
 */
export function reservationsFromDto(dtos: unknown[] | null | undefined): Reservation[] {
  return safeMap(dtos, reservationFromDto);
}

/**
 * Maps a UTXO DTO to a ReservationUtxo domain model.
 *
 * @param dto - The DTO from the OpenAPI client
 * @returns The ReservationUtxo domain model, or undefined if dto is null/undefined
 */
export function reservationUtxoFromDto(dto: unknown): ReservationUtxo | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    hash: safeString(d.hash),
    outputIndex: safeInt(d.outputIndex ?? d.output_index),
    script: safeString(d.script),
    value: safeString(d.value),
    blockHeight: safeString(d.blockHeight ?? d.block_height),
    reservedByRequestId: safeString(d.reservedByRequestId ?? d.reserved_by_request_id),
    reservationId: safeString(d.reservationId ?? d.reservation_id),
    valueString: safeString(d.valueString ?? d.value_string),
  };
}

File: taurus-protect-sdk-go/pkg/protect/mapper/reservation.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// ReservationFromDTO converts an OpenAPI Reservation to a domain Reservation.
func ReservationFromDTO(dto *openapi.TgvalidatordReservation) *model.Reservation {
	if dto == nil {
		return nil
	}

	reservation := &model.Reservation{
		ID:           safeString(dto.Id),
		Kind:         safeString(dto.Kind),
		Comment:      safeString(dto.Comment),
		AddressID:    safeString(dto.Addressid),
		Address:      safeString(dto.Address),
		Amount:       safeString(dto.Amount),
		ResourceID:   safeString(dto.ResourceId),
		ResourceType: safeString(dto.ResourceType),
	}

	// Convert creation date
	if dto.CreationDate != nil {
		reservation.CreationDate = *dto.CreationDate
	}

	// Convert currency info
	if dto.CurrencyInfo != nil {
		reservation.CurrencyInfo = CurrencyInfoFromDTO(dto.CurrencyInfo)
	}

	// Convert asset
	if dto.Asset != nil {
		reservation.Asset = AssetFromDTO(dto.Asset)
	}

	return reservation
}

// ReservationsFromDTO converts a slice of OpenAPI Reservation to domain Reservations.
func ReservationsFromDTO(dtos []openapi.TgvalidatordReservation) []*model.Reservation {
	if dtos == nil {
		return nil
	}
	reservations := make([]*model.Reservation, len(dtos))
	for i := range dtos {
		reservations[i] = ReservationFromDTO(&dtos[i])
	}
	return reservations
}

// ReservationUTXOFromDTO converts an OpenAPI UTXO to a domain ReservationUTXO.
func ReservationUTXOFromDTO(dto *openapi.TgvalidatordUTXO) *model.ReservationUTXO {
	if dto == nil {
		return nil
	}

	return &model.ReservationUTXO{
		ID:            safeString(dto.Id),
		Hash:          safeString(dto.Hash),
		OutputIndex:   safeInt64(dto.OutputIndex),
		Script:        safeString(dto.Script),
		Value:         safeString(dto.Value),
		ValueString:   safeString(dto.ValueString),
		BlockHeight:   safeString(dto.BlockHeight),
		ReservationID: safeString(dto.ReservationId),
	}
}
